flink-tests/src/test/java/org/apache/flink/test/runtime/SchedulingLocalRecoveryITCase.java

```diff
@@ -0,0 +1,227 @@
+            CompletableFuture<Void> terminationFuture = miniCluster.terminateTaskManager(0);
+            terminationFuture.get();
+
+            // restart a taskmanager as a replacement for the killed one
+            miniCluster.startTaskManager();
+
+            // wait until all tasks running again
+            waitUntilAllVerticesRunning(jobId, miniCluster, TIMEOUT);
+
+            ArchivedExecutionGraph graph =
+                    miniCluster.getArchivedExecutionGraph(jobGraph.getJobID()).get();
+
+            miniCluster.cancelJob(jobId).get();
+
+            return graph;
+        }
+    }
+
+    private void waitUntilAllVerticesRunning(
+            JobID jobId, MiniCluster miniCluster, long maxWaitMillis) throws Exception {
+        final Deadline deadline = Deadline.fromNow(Duration.ofMillis(maxWaitMillis));
+
+        boolean checkResult = false;
+        do {
+            AccessExecutionGraph graph = miniCluster.getArchivedExecutionGraph(jobId).get();
+
+            // we need the job status to be RUNNING first to ensure execution vertices are created
+            if (graph.getState() != JobStatus.RUNNING) {
+                continue;
+            }
+
+            checkResult = areAllVerticesRunning(graph);
+
+            if (!checkResult) {
+                Thread.sleep(2L);
+            }
+        } while (!checkResult && deadline.hasTimeLeft());
+
+        if (!checkResult) {
+            throw new TimeoutException("Not all vertices becomes RUNNING in time.");
+        }
+    }
+
+    private boolean areAllVerticesRunning(AccessExecutionGraph graph) {
+        for (AccessExecutionVertex vertex : graph.getAllExecutionVertices()) {
+            if (vertex.getExecutionState() != ExecutionState.RUNNING) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private JobGraph createJobGraph(int parallelism) throws IOException {
+        final JobVertex source = new JobVertex("v1");
+        source.setInvokableClass(PendingInvokable.class);
+        source.setParallelism(parallelism);
+
+        ExecutionConfig executionConfig = new ExecutionConfig();
+        executionConfig.setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 10));
+
+        return JobGraphBuilder.newStreamingJobGraphBuilder()
+                .addJobVertices(Arrays.asList(source))
+                .setExecutionConfig(executionConfig)
+                .build();
+    }
+
+    /** Invokable which does not finish. */
+    public static final class PendingInvokable extends AbstractInvokable {
+        private boolean canceled = false;
+
+        public PendingInvokable(Environment environment) {
+            super(environment);
+        }
+
+        @Override
+        public void invoke() throws Exception {
+            while (!canceled) {
+                Thread.sleep(100);
+            }
+        }
+
+        @Override
+        public void cancel() {
+            canceled = true;
+        }
+    }
```

Comment:
This looks quite similar to `MyCancellableInvokable`. Maybe this can deduplicated.

Reply:
Yes. I will factor `MyCancellableInvokable` out to be a standalone test class in `flink-runtime`.

Reply:
Done via 39fd333 and d0d0ab6